manpages/clip.man
.na
.nh
.TH CLIP 1 3.4.18 BL3DEMC
.SH NAME
clip \- command line image processing for mrc files.
.SH SYNOPSIS
clip process  [options]  input_files...  [output_file]
.SH DESCRIPTION
The processes that clip can do are listed below. One and only one
process must be selected on the command line.  Only the first three or
four letters of the process name have to be entered.
.TP
.I add
Add several images together.
.TP
.I average (avg) 
Average several images together.
.PD 0
.TP
.I brighten
Brighten image by scaling intensities.
.TP
.I color
Add false color to image.
.TP
.I contrast
Adjust contrast by scaling intensities.
.TP
.I correlation   
Do a auto/cross correlation.
.TP
.I diffusion
Do 2d anisotropic diffusion.
.TP
.I divide
Divide one image volume by another or by a single image.
.TP
.I gradient
Calculate a gradient as in 3dmod image processing.
.TP
.I graham
Apply Graham filter as in 3dmod image processing.
.TP
.I info
Print header information to stdout.
.TP   
.I fft
Calculate a FFT or inverse FFT transform.
.TP
.I filter
Calculate a bandpass filter.
.TP
.I flip[xyz]
Flip an image by x y or z.
.TP
.I histogram
Print histogram of values.
.TP
.I joinrgb
Join images from 3 byte files into one RGB file.
.TP
.I laplacian
Apply Laplacian filter as in 3dmod image processing.
.TP
.I median
Apply median filter.
.TP
.I multiply
Multiply one image volume by another or by a single image.
.TP
.I prewitt
Apply Prewitt filter as in 3dmod image processing.
.TP
.I quadrant
Correct disparities between quadrants in images from 4-port readout camera.
.TP
.I resize
Box out image to a new size.
.TP
.I rotx
Rotate a volume by -90 degrees about X axis.
.TP
.I shadow
Adjust darkness of image by scaling intensities.
.TP
.I sharpen
Sharpen image as in 3dmod image processing.
.TP
.I smooth
Smooth image as in 3dmod image processing.
.TP
.I sobel
Apply Sobel filter as in 3dmod image processing.
.TP
.I splitrgb
Split an RGB file into 3 byte files.
.TP
.I stats
Print min, max, mean, standard deviation, and location of min and max.
.TP
.I standev
Compute the standard deviation of a set of images.
.TP
.I subtract
Subtract one image volume from another.
.TP
.I truncate
Limit pixel values at low or high end, or both.
.TP
.I unpack
Unpack 4-bit values packed into bytes, with optional scaling by reference.
.TP
.I unwrap
Undo wraparound of values in integer data.
.TP
.I variance
Compute the variance of a set of images.
.PD
.SH Options
These options are available to most processes.
.TP
.B \-v
View output data file using 3dmod.
.TP
.B \-2d
Use 2d instead of 3d (default) processes if a
2d process exists. Each input file is considered
a stack of 2D images to be processed.  For most processes, this will change
only the way that the 
.B -iz
option is interpreted.
.TP
.B \-3d
Use 3d process. Each input file is considered a 
volume to be processed.
.TP
.B -a
Append data to output file.  Append and overwrite are not available for
processing modes that do not take standard input and output size and centering
options.
.TP
.B -ov \fIsection\fR
Overwrite output starting at section number
.I section.
Range is from 0 (first section) to z size - 1 (last section) unless the
.B -1
option is entered to number sections from 1.
.TP
.B \-p \fIvalue\fR
Image coordinates with no image data are padded with the given 
.I value.
The default is the mean value of the input image.
.TP
.B -c[xyz] \fIvalue\fR
Adjust the center of input image. Default is center of first input 
file.
The
.B -cx 
option will center the x coordinate on the given 
.I value,
the 
.B -cy
option will center the y coordinate on the given
.I value and
the
.B -cz 
option will center the z coordinate on the given
.I value.

.TP
.B -ix \fIvalue\fR
.PD 0
.TP
.B -iy \fIvalue\fR
.TP
.B -iz \fIvalue\fR
.PD
Sets the size of input image.  
The default is the size of the first input file.
The 
.B -ix
option sets the x input size,
the
.B -iy
option sets the y input size.  The
.B -iz
option sets the z input size, or specifies a list of Z values if the
.B -2d
option is included.  In the latter case, the 
.I value
is interpreted as a comma-separated list of ranges. Commas indicate individual
section numbers and dashes (or minus symbol) indicate a range of sections.
.TP
.B -x \fIvalue,value\fR
.PD 0
.TP
.B -y \fIvalue,value\fR
.PD
These options are an alternative way of specifying the size and center of the 
input image.  The two values are the starting and ending coordinates,
numbered from 0 (i.e., the first pixel in an image is (0, 0)).  The
.B -x
option sets the X coordinates, and neither 
.B -cx
nor
.B -ix
may be entered with it.  The
.B -y
option sets the Y coordinates, and neither 
.B -cy
nor
.B -iy
may be entered with it.  These coordinates are not constrained to lie within
the volume; regions outside the volume will be filled with the padding value.
.TP
.B -ox \fIvalue\fR
.PD 0
.TP
.B -oy \fIvalue\fR
.TP
.B -oz \fIvalue\fR
.PD
Size of output file.  Default is same as input size, 
options 
.B -a 
and 
.B -ov 
override these setting.
The 
.B -ox 
option sets the x output size,
the 
.B -oy
options sets the y output size and the
.B -oz
option sets the z output size.
.TP
.B -1
Z values are numbered from 1 instead of 0, the default.  This option affects
the interpretation of lists of Z values entered with 
.B -iz
when the 
.B -2d
option is given, and the interpretation of the Z value entered with the
.B -oz
and
.B -cz
options.  When running "clip stat", slices will be referred to as views
instead and numbered from 1 with this option.

.SH Selected options
The following options are valid for selected processes:
.TP
.B -s
Switch, use depends on process.
.TP
.B -m \fImode\fR
Output modes: "byte", "ubyte", "sbyte", "short", "float", "complex", "ushort",
"rgb", or 0-4, 6, or 16.  0 = byte, 1 = short, 2 = float, 3 = complex short, 
4 = complex float, 6 = unsigned short, 16 = rgb.  "byte" will produce byte
output that is signed or unsigned depending on the default for this version of
IMOD and the value of the environment variable WRITE_MODE0_SIGNED.  Use
"ubyte" or "sbyte" to force unsigned or signed byte output regardless of other
settings.  Unsigned output is needed to work with versions of IMOD
before IMOD 4.2.23; signed output may be needed for compatibility with
external software that has followed recent documentation of the MRC format.
Regardless of the representation in the file, bytes are read into IMOD programs as
unsigned with a range of 0 to 255.
.TP
.B -h \fIlevel\fR
Has five uses: 1) Level for high pass filter. Range is (0.0 - 0.71 / pixel).
The default value is 0. 2) High level for image truncation; the default is no
truncation.  3) Width of area analyzed in quadrant correction.  4) Upper
limit for histogram output.  5) Threshold above

History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)


clip/clip.c
/*
 *  clip -- Command Line Image Proccesing
 *
 *  Original author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 */

#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include "mrcfiles.h"
#include "mrcslice.h"
#include "clip.h"
#include "imodconfig.h"
#include "b3dutil.h"

void usage(void)
{
  char *name = "clip";

  fprintf(stderr, 
          "%s: Command Line Image Processing. %s, %s %s\n",
          name,
          VERSION_NAME, __DATE__, __TIME__);
  imodCopyright();
  fprintf(stderr, "----------------------------------------------------\n");

  fprintf(stderr, "%s usage:\n", name);
  fprintf(stderr, "%s [process] [options] [input files...] [output file]\n", name);
  fprintf(stderr, "\nprocess:\n");
  fprintf(stderr, "\tadd         - Add images together.\n");
  fprintf(stderr, "\taverage     - Average files together.\n");
  fprintf(stderr, "\tbrightness  - Increase or decrease brightness.\n");
  fprintf(stderr, "\tcolor       - Add false color.\n");
  fprintf(stderr, "\tcontrast    - Increase or decrease contrast.\n");
  fprintf(stderr, "\tcorrelation - Do a auto/cross correlation.\n");
  fprintf(stderr, "\tdiffusion   - Do 2-D anisotropic diffusion on slices.\n");
  fprintf(stderr, "\tdivide      - Divide one image volume by another.\n");
  fprintf(stderr, "\tgradient    - Compute gradient as in 3dmod.\n");
  fprintf(stderr, "\tgraham      - Apply Graham filter as in 3dmod.\n");
  fprintf(stderr, "\thistogram   - Print histogram of values.\n");
  fprintf(stderr, "\tinfo        - Print information to stdout.\n");
  fprintf(stderr, "\tfft         - Do a fft or inverse fft transform.\n");
  fprintf(stderr, "\tfilter      - Do a bandpass filter.\n");
  fprintf(stderr, "\tflip..      - Flip image about various axes.\n");
  /*     fprintf(stderr, "\tpeak        - Find peaks in image.\n"); */
  fprintf(stderr, "\tjoinrgb     - Join 3 byte files into an RGB file.\n");
  fprintf(stderr, "\tlaplacian   - Apply Laplacian filter as in 3dmod.\n");
  fprintf(stderr, "\tmedian      - Do median filtering.\n");
  fprintf(stderr, "\tmultiply    - Multiple one image volume by another.\n");
  fprintf(stderr, "\tprewitt     - Apply Prewitt filter as in 3dmod.\n");
  fprintf(stderr, "\tquadrant    - Correct quadrant disparities from 4-port camera.\n");
  fprintf(stderr, "\tresize      - Cut out and/or pad image data.\n");
  fprintf(stderr, "\trotx        - Rotate volume by -90 about X axis.\n");
  /* fprintf(stderr, "\trotation    - Rotate image\n"); */
  fprintf(stderr, "\tshadow      - Increase or decrease image shadows.\n");
  fprintf(stderr, "\tsharpen     - Sharpen image as in 3dmod.\n");
  fprintf(stderr, "\tsmooth      - Smooth image as in 3dmod.\n");
  fprintf(stderr, "\tsobel       - Apply Sobel filter as in 3dmod.\n");
  fprintf(stderr, "\tsplitrgb    - Split RGB image file into 3 byte files.\n");
  fprintf(stderr, "\tstandev     - Compute standard deviation for averaged images.\n");
  fprintf(stderr, "\tstats       - Print some stats on image file.\n");
  fprintf(stderr, "\tsubtract    - Subtract one image volume from another.\n");
  fprintf(stderr, "\ttruncate    - Limit image values at low or high end.\n");
  fprintf(stderr, "\tunwrap      - Undo a wraparound of integer intensity values.\n");
  fprintf(stderr, "\tunpack      - Unpack 4-bit data into bytes [& multiply by gain "
          "ref].\n");
  fprintf(stderr, "\tvariance    - Compute variance for averaged images.\n");
  /* fprintf(stderr, "\ttranslate   - translate image.\n");
  fprintf(stderr, "\tzoom        - magnify image.\n"); */
  fprintf(stderr, "\noptions:\n");
  fprintf(stderr, "\t[-v]  view output data.\n");
  fprintf(stderr, "\t[-3d] or [-2d] treat image as 3d (default) or 2d.\n");
  fprintf(stderr, "\t[-s] Switch, [-n #] Amount; Depends on function.\n");
  fprintf(stderr, "\t[-n #] [-l #] Iterations and Gaussian sigma for smoothing.\n");
  fprintf(stderr, "\t[-h #] [-l #] values for high/low pass filters or truncation.\n");
  fprintf(stderr, "\t[-cc #] [-l #] [-k #] values for anisotropic diffusion.\n");
  fprintf(stderr, "\t[-r #] [-g #] [-b #] red, green, blue values.\n");
  fprintf(stderr, "\t[-x #,#]  [-y #,#]  starting and ending input coords.\n");
  fprintf(stderr, "\t[-cx #]  [-cy #]  [-cz #]  center coords.\n");
  fprintf(stderr, "\t[-ix #]  [-iy #]  [-iz #]  input sizes.\n");
  fprintf(stderr, "\t[-ox #]  [-oy #]  [-oz #]  output sizes.\n");
  fprintf(stderr, "\t[-a] Append output to file.\n");
  fprintf(stderr, "\t[-ov #] Overwrite output starting at section #\n");
  fprintf(stderr, "\t[-m (mode#)] Output data mode.\n");
  fprintf(stderr, "\t[-p (pad#)] Pad empty data value.\n");
  fprintf(stderr, "\t[-1] Number Z values from 1 instead of 0.\n");
  fprintf(stderr, "\t[-P file] Name of piece list file for stats on a montage.\n");
  fprintf(stderr, "\t[-O #,#]  Overlaps in X and Y in displayed montage.\n");
  fprintf(stderr, "\n");
}

/* 11/4/04: switch to standard out and standard format to some extent */
void show_error(const char *format, ...)
{
  char errorMess[512];
  va_list args;
  va_start(args, format);
  vsprintf(errorMess, format, args);
  printf("ERROR: %s\n", errorMess);
}

void show_warning(char *reason)
{
  printf("WARNING: %s\n", reason);
}

void show_status(char *info)
{
  printf("%s", info);
  fflush(stdout);
}

void default_options(ClipOptions *opt)
{
  opt->hin = opt->hin2 = opt->hout = NULL;
  opt->x  = IP_DEFAULT; opt->x2 = IP_DEFAULT;
  opt->y  = IP_DEFAULT; opt->y2 = IP_DEFAULT;
  opt->z  = IP_DEFAULT; opt->z2 = IP_DEFAULT;
  opt->ix = IP_DEFAULT; opt->iy = IP_DEFAULT; 
  opt->iz = IP_DEFAULT; opt->iz2 = IP_DEFAULT;
  opt->ox = IP_DEFAULT; opt->oy = IP_DEFAULT; opt->oz = IP_DEFAULT;
  opt->cx = IP_DEFAULT; opt->cy = IP_DEFAULT; opt->cz = IP_DEFAULT;
  opt->outBefore = opt->outAfter = IP_DEFAULT;
  opt->red = IP_DEFAULT; opt->green = IP_DEFAULT; opt->blue = IP_DEFAULT;
  opt->high = IP_DEFAULT; opt->low = IP_DEFAULT;
  opt->thresh = IP_DEFAULT;
  opt->weight = IP_DEFAULT;
  opt->pad    = IP_DEFAULT;
  opt->process = IP_NONE;
  opt->dim = 3;
  opt->add2file = IP_APPEND_FALSE;
  opt->sano = FALSE;
  opt->val = IP_DEFAULT;
  opt->mode = IP_DEFAULT;
  opt->nofsecs = IP_DEFAULT;
  opt->secs = NULL;
  opt->ocanresize = TRUE;
  opt->ocanchmode = TRUE;
  opt->fromOne = FALSE;
  opt->ofname = NULL;
  opt->plname = NULL;
  opt->newXoverlap = IP_DEFAULT;
  opt->newYoverlap = IP_DEFAULT;
}



int main( int argc, char *argv[] )
{
  MrcHeader hin, hin2, hout;
  ClipOptions opt;

  int process = IP_NONE;  /* command to run.             */
  int view    = FALSE;    /* view file at end?           */
  int procout = TRUE;     /* will process write output?. */
  int needtwo = FALSE;    /* Does process need two input files? */
  int i, j;
  int retval = 0;

  char viewcmd[1024];
  char *progname = imodProgName(argv[0]);

  if (argc < 3){
    usage();
    exit(3);
  }

  opt.command = argv[1];

  /* Find which process to run */
  if (!strncmp( argv[1], "add", 3)) {
    process = IP_ADD;
    needtwo = TRUE;
  }
  if ((!strncmp( argv[1], "avg", 3)) || 
      (!strncmp( argv[1], "average", 3)) )
    process = IP_AVERAGE;
  if (!strncmp( argv[1], "standev", 4))
    process = IP_STANDEV;
  if (!strncmp( argv[1], "variance", 3))
    process = IP_VARIANCE;
  if (!strncmp( argv[1], "multiply", 3)) {
    process = IP_MULTIPLY;
    needtwo = TRUE;
  }
  if (!strncmp( argv[1], "subtract", 3)) {
    process = IP_SUBTRACT;
    needtwo = TRUE;
  }
  if (!strncmp( argv[1], "divide", 3)) {
    process = IP_DIVIDE;
    needtwo = TRUE;
  }
  if (!strncmp( argv[1], "brightness", 3))
    process = IP_BRIGHTNESS;
  if (!strncmp( argv[1], "color", 3))
    process = IP_COLOR;
  if (!strncmp( argv[1], "contrast", 3))
    process = IP_CONTRAST;

  if (!strncmp( argv[1], "correlation", 3))
    process = IP_CORRELATE; 
  if (!strncmp( argv[1], "diffusion", 3))
    process = IP_DIFFUSION; 

  if (!strncmp( argv[1], "info", 3)){
    process = IP_INFO;
    procout = FALSE;
  }

  if (!strncmp( argv[1], "fft", 3))
    process = IP_FFT;
  if (!strncmp( argv[1], "filter", 3))
    process = IP_FILTER;

  if (!strncmp( argv[1], "flip", 4))
    process = IP_FLIP;
  if (!strncmp( argv[1], "rotx", 4))
    process = IP_FLIP;
  if (!strncmp( argv[1], "gradient", 4))
    process = IP_GRADIENT;
  if (!strncmp( argv[1], "graham", 4))
    process = IP_GRAHAM;
  if (!strncmp( argv[1], "histogram", 2)) {
    process = IP_HISTOGRAM;
    procout = FALSE;
  }
  if (!strncmp( argv[1], "laplacian", 2))
    process = IP_LAPLACIAN;
  if (!strncmp( argv[1], "median", 2))
    process = IP_MEDIAN;
  /*  if (!strncmp( argv[1], "peak", 3)){
    process = IP_PEAK;
    procout = FALSE;
    } */
  if (!strncmp( argv[1], "prewitt", 2))
    process = IP_PREWITT;
  if (!strncmp( argv[1], "resize", 3))
    process = IP_RESIZE;
  /* if (!strncmp( argv[1], "rotate", 3))
     process = IP_ROTATE; */
  if (!strncmp( argv[1], "shadow", 4))
    process = IP_SHADOW;
  if (!strncmp( argv[1], "sharpen", 4))
    process = IP_SHARPEN;
  if (!strncmp( argv[1], "smooth", 2))
    process = IP_SMOOTH;
  if (!strncmp( argv[1], "sobel", 2))
    process = IP_SOBEL;
  if (!strncmp( argv[1], "stat", 4)){
    process = IP_STAT;
    procout = FALSE;
  }
  /* if (!strncmp( argv[1], "translate", 2))
    process = IP_TRANSLATE;
  if (!strncmp( argv[1], "zoom", 3))
     process = IP_ZOOM;
  if (!strncmp( argv[1], "project", 3)){
    process = IP_PROJECT;
    } */

  if (!strncmp( argv[1], "truncate", 3))
    process = IP_TRUNCATE;
  if (!strncmp( argv[1], "unwrap", 3))
    process = IP_UNWRAP;
  if (!strncmp( argv[1], "quadrant", 2)) {
    process = IP_QUADRANT;
    opt.dim = 2;
  } 
  if (!strncmp( argv[1], "unpack", 3))
    process = IP_UNPACK;
  if (!strncmp( argv[1], "splitrgb", 3)){
    process = IP_SPLITRGB;
  }
  if (!strncmp( argv[1], "joinrgb", 3)){
    process = IP_JOINRGB;
  }

  if (process == IP_NONE){
    usage();
    exit(1);
  }

  default_options(&opt);
  opt.process = process;
  opt.pname = progname;

  /* get options */
  for (i = 2; i < argc; i++)
    if (argv[i][0] == '-')
      switch (argv[i][1]){
		    
      case 'a':
        opt.add2file = IP_APPEND_ADD;  break;

      case '3':
        if (process != IP_QUADRANT)
          opt.dim = 3;
        break;

      case '2':
        opt.dim = 2; break;

      case '1':
        opt.fromOne = TRUE; break;

      case 's':
        opt.sano = TRUE; break;		    
		    
      case 'n':
        sscanf(argv[++i], "%f", &(opt.val)); break;
		    
      case 'm':
        opt.mode = sliceMode(argv[++i]); 
        if (opt.mode == SLICE_MODE_UNDEFINED) {
          fprintf(stderr, "%s: invalid mode entry %s.\n", progname, argv[i]);
          exit(3);
        }
        if (opt.mode == SLICE_MODE_SBYTE || opt.mode == SLICE_MODE_UBYTE) {
          overrideWriteBytes(opt.mode == SLICE_MODE_SBYTE ? 1 : 0);
          opt.mode = 0;
        }
        break;

      case 'v':
        view = TRUE; break;

      case 'p':  /* Pad option */
        opt.pad = (float)atof(argv[++i]); break;

      case 't':
        sscanf(argv[++i], "%f", &(opt.thresh)); break;

      case 'k':
      case 'w':
        sscanf(argv[++i], "%f", &(opt.weight)); break;

      case 'r':
        sscanf(argv[++i], "%f", &(opt.red)); break;
      case 'g':
        sscanf(argv[++i], "%f", &(opt.green)); break;
      case 'b':
        sscanf(argv[++i], "%f", &(opt.blue)); break;

      case 'l':
        sscanf(argv[++i], "%f", &(opt.low)); break;
      case 'h':
        sscanf(argv[++i], "%f", &(opt.high)); break;

      case 'x': case 'X':
        sscanf(argv[++i], "%d%*c%d", &(opt.x), &(opt.x2)); break;
      case 'y': case 'Y':
        sscanf(argv[++i], "%d%*c%d", &(opt.y), &(opt.y2)); break;
        /*case 'z': case 'Z':
          sscanf(argv[++i], "%f%*c%f", &(opt.z), &(opt.z2)); break; */

      case 'o':
        switch (argv[i][2]){
        case 0x00:
        case ' ':
        case 'v':
          opt.add2file = IP_APPEND_OVERWRITE;
          sscanf(argv[++i], "%d", &(opt.isec));
          break;
        case 'x': case 'X':
          sscanf(argv[++i], "%d", &(opt.ox)); break;
        case 'y': case 'Y':
          sscanf(argv[++i], "%d", &(opt.oy)); break;
        case 'z': case 'Z':
          sscanf(argv[++i], "%d", &(opt.oz)); break;
        default:
          fprintf(stderr, "%s: invalid option %s.\n", progname, argv[i]);
          exit(3);
        }
        break;
		    
      case 'i': case 'I':
        switch (argv[i][2]){
        case 'x': case 'X':
          sscanf(argv[++i], "%d", &(opt.ix)); break;
        case 'y': case 'Y':
          sscanf(argv[++i], "%d", &(opt.iy)); break;
        case 'z': case 'Z':
          sscanf(argv[++i], "%d%*c%d", &(opt.iz),&opt.iz2);
          opt.secs = clipMakeSecList(argv[i], &opt.nofsecs);
          break;
        default:
          fprintf(stderr, "%s: invalid option %s.\n", progname, argv[i]);
          exit(3);
        }
        break;

      case 'c': case 'C':
        switch (argv[i][2]){
        case 'x': case 'X':
          sscanf(argv[++i], "%g", &(opt.cx)); break;
        case 'y': case 'Y':
          sscanf(argv[++i], "%g", &(opt.cy)); break;
        case 'z': case 'Z':
          sscanf(argv[++i], "%g", &(opt.cz)); break;
        case 'c': case 'C':
          sscanf(argv[++i], "%f", &(opt.thresh)); break;
        default:
          fprintf(stderr, "%s: invalid option %s.\n", progname, argv[i]);
          exit(3);
        }
        break;

      case 'P':
        opt.plname = strdup(argv[++i]);
        break;

      case 'O':
        sscanf(argv[++i], "%d%*c%d", &opt.newXoverlap, &opt.newYoverlap);
        break;

      default:
        fprintf(stderr, "%s: invalid option %s.\n", progname, argv[i]);
        exit(3);
      }
    else
      break;


  opt.fnames  = &(argv[i]);

  if (!procout) {
    /* check for at least one input file */
    if ((argc - 1) < i){
      usage();
      exit(3);
    }
    opt.infiles = argc - i;
  } else {
    /* check for at least one input and one output file name. */
    if ((argc - 2) < i){
      usage();
      exit(3);
    }
    opt.infiles = argc - i - 1;
  }

  if (opt.x != IP_DEFAULT) {
    if (opt.cx != IP_DEFAULT || opt.ix != IP_DEFAULT) {
      fprintf(stderr, "You cannot use -x together with -cx or -ix\n");
      exit(1);
    }
  }
  if (opt.y != IP_DEFAULT) {
    if (opt.cy != IP_DEFAULT || opt.iy != IP_DEFAULT) {
      fprintf(stderr, "You cannot use -y together with -cy or -iy\n");
      exit(1);
    }
  }


  /* check and load files:
   * Always at least one input file.
   */
  hin.fp = fopen(argv[i], "rb");
  hin.pathname = argv[i];
  if (!hin.fp){
    fprintf(stderr, "Error opening %s\n", argv[i]);
    return(-1);
  }
  if (mrc_head_read(hin.fp, &hin)){
    fprintf(stderr, "Error reading %s\n", argv[i]);
    return(-1);
  }

  /* Set output header default same as first input file. */
  hout = hin;

  /* 7/20/11: Consolidate all changes for output header into this call; eliminate 
   * setting swapped to 0 below; this header is replaced if appending
   * There is also a new header made in lots of places, not sure how much this is used */
  mrcInitOutputHeader(&hout);

  /* Load additional input files. */
  i++;
  if (opt.infiles > 1){
    hin2.fp = fopen(argv[i], "rb");
    hin2.pathname = argv[i];
    if (!hin2.fp){
      fprintf(stderr, "Error opening %s\n", argv[i]);
      return(-1);
    }
    if (mrc_head_read(hin2.fp, &hin2)){
      if (process == IP_INFO){
        mrc_head_print(&hin);
        printf("**********************************************\n");
        printf("WARNING: This file is not a readable MRC file.\n");
        printf("**********************************************\n");
      }
      fprintf(stderr, "Error reading %s\n", argv[i]);
      return(-1);
    }
    i++;
  }


  /* Setup output file if needed and there are enough - otherwise let the process give 
     the error message . */
  if (procout && (!needtwo || opt.infiles > 1)){
    opt.ofname = argv[argc - 1];
    if (opt.add2file){
      hout.fp = fopen(argv[argc - 1], "rb+");
      if (!hout.fp){
        fprintf(stderr, "Error finding %s\n", argv[argc - 1]);
        return(-1);
      }

      if (mrc_head_read(hout.fp, &hout)){
        fprintf(stderr, "Error reading %s\n", argv[i]);
        return(-1);
      }

    } else if (process != IP_SPLITRGB) {

      /* DNm 10/20/03: switch to calling routine for backup file */
      if (!getenv("IMOD_NO_IMAGE_BACKUP"))
        imodBackupFile(argv[argc - 1]);

      hout.fp = fopen(argv[argc - 1], "wb+");
        
      if (!hout.fp){
        fprintf(stderr, "Error opening %s\n", argv[argc - 1]);
        return(-1);
      }
    }
  }

  opt.hin = &hin;
  opt.hout = &hout;

  /* Massage the Z values if numbered from one */
  if (opt.fromOne) {
    if (opt.add2file == IP_APPEND_OVERWRITE)
      opt.isec--;
    if (opt.nofsecs != IP_DEFAULT)
      for (j = 0; j < opt.nofsecs; j++)
        opt.secs[j]--;
    if (opt.cz != IP_DEFAULT)
        opt.cz--;
    if (opt.dim == 2) {
      if (opt.iz != IP_DEFAULT)
        opt.iz--;
      if (opt.iz2 != IP_DEFAULT)
        opt.iz2--;
    }
  }

  /* run the selected process */
  switch(process){
  case IP_AVERAGE:
  case IP_STANDEV:
  case IP_VARIANCE:
  case IP_ADD:
  case IP_SUBTRACT:
    retval = clip_average(&hin, &hin2, &hout, &opt);
    break;
  case IP_MULTIPLY:
  case IP_DIVIDE:
    retval = clip_multdiv(&hin, &hin2, &hout, &opt);
    break;
  case IP_UNPACK:
    retval = clipUnpack(&hin, &hin2, &hout, &opt);
    break;
  case IP_BRIGHTNESS:
  case IP_CONTRAST:
  case IP_SHADOW:
  case IP_RESIZE:
  case IP_TRUNCATE:
  case IP_UNWRAP:
    retval = clip_scaling(&hin, &hout, &opt);
    break;
  case IP_COLOR:
    retval = clip_color(&hin, &hout, &opt);
    break;
  case IP_QUADRANT:
    retval = clip_quadrant(&hin, &hout, &opt);
    break;
  case IP_CORRELATE:
    retval = grap_corr(&hin, &hin2, &hout, &opt);
    break;
  case IP_DIFFUSION:
    retval = clipDiffusion(&hin, &hout, &opt);
    break;
  case IP_GRADIENT:
  case IP_GRAHAM:
  case IP_PREWITT:
  case IP_SOBEL:
    retval = clipEdge(&hin, &hout, &opt);
    break;
  case IP_INFO:
    retval = mrc_head_print(&hin);
    break;
  case IP_FFT:
    retval = clip_fft(&hin, &hout, &opt);
    break;
  case IP_FILTER:
    /* opt.dim = 2; */
    retval = clip_bandpass_filter(&hin, &hout, &opt);
    break;
  case IP_FLIP:
    retval = clip_flip(&hin, &hout, &opt);
    break;
  case IP_HISTOGRAM:
    retval = clipHistogram(&hin, &opt);
    break;
  case IP_JOINRGB:
    retval = clip_joinrgb(&hin, &hin2, &hout, &opt);
    break;
  case IP_LAPLACIAN:
  case IP_SMOOTH:
  case IP_SHARPEN:
    retval = clip_convolve(&hin, &hout, &opt);
    break;
  case IP_MEDIAN:
    retval = clipMedian(&hin, &hout, &opt);
    break;
  case IP_PEAK:
    retval = puts("peak: future function\n");
    break;
  case IP_ROTATE:
    retval = grap_rotate(&hin, &hout, &opt);
    break;
  case IP_SPLITRGB:
    retval = clip_splitrgb(&hin, &opt);
    break;
  case IP_STAT:
    retval = clip_stat(&hin, &opt);
    break;
  case IP_TRANSLATE:
    retval = grap_trans(&hin, &hout, &opt);
    break;
  case IP_ZOOM:
    if (grap_zoom(&hin, &hout, &opt)) {
      fprintf(stderr, "%s: error doing zoom.\n", progname);
      retval = 1;
    }
    break;

  default:
    fprintf(stderr, "%s error: no process selected.\n", progname);
    retval = 1;
  }

  if (retval)
    exit(retval);
  fclose (hin.fp);
  if (procout && process != IP_SPLITRGB)
    fclose (hout.fp);

  if (view){
    sprintf(viewcmd, "3dmod %s", argv[argc - 1]);
    system(viewcmd);
  }
  return(0);
}
     
/*
 * Make a list of sections for use with the 
 * -2d and -z option. commands can be like
 * -z 0,4-10 or -z 2,3,4-8,12,19
 */
int *clipMakeSecList(char *clst, int *nofsecs)
{
  int len;
  int i;
  int *secs;
  int *tsecs;
  int range, top, r;

  secs = (int *)malloc(sizeof(int));
  len = strlen(clst);
  secs[0] = atoi(clst);
  *nofsecs = 1;
     
  for(i = 0; i < len; i++){
    switch(clst[i]){
    case ',':
      (*nofsecs)++;
      tsecs = (int *)realloc(secs, sizeof(int) * (*nofsecs));
      secs = tsecs;
      secs[*nofsecs - 1] = atoi(&(clst[++i]));
      break;
	       
    case '-':
      top = atoi(&(clst[++i]));
      if (top < secs[*nofsecs - 1]){
        range = top;
        top = secs[*nofsecs - 1];
        secs[*nofsecs - 1] = range;
      }
      if (top == secs[*nofsecs - 1])
        continue;
      range = top - secs[*nofsecs - 1];
      tsecs = (int *)realloc(secs, sizeof(int) * (*nofsecs + range));
      secs = tsecs;
      for (r = 0; r < range; r++)
        secs[*nofsecs + r] = secs[*nofsecs - 1] + r + 1;
      *nofsecs += range;
      break;
	       
    default:
      break;
    }
  }
  return(secs);
}

